File: rust/src/models/invoice_event_data.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{customer_id::CustomerId, invoice_id::InvoiceId, invoice_status::InvoiceStatus};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct InvoiceEventData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub consolidated_into_invoice_id: Option<InvoiceId>,

    pub created_at: String,

    pub currency: String,

    /// User-defined custom property values, keyed by definition key.
    pub custom_properties: serde_json::Value,

    pub customer_id: CustomerId,

    pub invoice_id: InvoiceId,

    /// Absent while the invoice is a draft — the number is assigned at finalization.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invoice_number: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_invoice_id: Option<InvoiceId>,

    pub status: InvoiceStatus,

    pub tax_amount: i32,

    pub total: i32,
}

impl InvoiceEventData {
    pub fn new(
        created_at: String,
        currency: String,
        custom_properties: serde_json::Value,
        customer_id: CustomerId,
        invoice_id: InvoiceId,
        status: InvoiceStatus,
        tax_amount: i32,
        total: i32,
    ) -> Self {
        Self {
            consolidated_into_invoice_id: None,
            created_at,
            currency,
            custom_properties,
            customer_id,
            invoice_id,
            invoice_number: None,
            parent_invoice_id: None,
            status,
            tax_amount,
            total,
        }
    }
}

File: typescript/src/models/invoiceEventData.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type CustomerId, CustomerIdSerializer } from "./customerId";
import { type InvoiceId, InvoiceIdSerializer } from "./invoiceId";
import { type InvoiceStatus, InvoiceStatusSerializer } from "./invoiceStatus";

export interface InvoiceEventData {
  consolidatedIntoInvoiceId?: InvoiceId | null;

  createdAt: Date;

  currency: string;

  /** User-defined custom property values, keyed by definition key. */
  customProperties: any;

  customerId: CustomerId;

  invoiceId: InvoiceId;

  /** Absent while the invoice is a draft — the number is assigned at finalization. */
  invoiceNumber?: string | null;

  parentInvoiceId?: InvoiceId | null;

  status: InvoiceStatus;

  taxAmount: number;

  total: number;
}

export const InvoiceEventDataSerializer = {
  _fromJsonObject(object: any): InvoiceEventData {
    return {
      consolidatedIntoInvoiceId:
        object["consolidated_into_invoice_id"] != null
          ? InvoiceIdSerializer._fromJsonObject(object["consolidated_into_invoice_id"])
          : undefined,
      createdAt: parseDateTime(object["created_at"]),
      currency: object["currency"],
      customProperties: object["custom_properties"],
      customerId: CustomerIdSerializer._fromJsonObject(object["customer_id"]),
      invoiceId: InvoiceIdSerializer._fromJsonObject(object["invoice_id"]),
      invoiceNumber: object["invoice_number"],
      parentInvoiceId:
        object["parent_invoice_id"] != null
          ? InvoiceIdSerializer._fromJsonObject(object["parent_invoice_id"])
          : undefined,
      status: InvoiceStatusSerializer._fromJsonObject(object["status"]),
      taxAmount: object["tax_amount"],
      total: object["total"],
    };
  },

  _toJsonObject(self: InvoiceEventData): any {
    return {
      consolidated_into_invoice_id:
        self.consolidatedIntoInvoiceId != null
          ? InvoiceIdSerializer._toJsonObject(self.consolidatedIntoInvoiceId)
          : undefined,
      created_at: self.createdAt,
      currency: self.currency,
      custom_properties: self.customProperties,
      customer_id: CustomerIdSerializer._toJsonObject(self.customerId),
      invoice_id: InvoiceIdSerializer._toJsonObject(self.invoiceId),
      invoice_number: self.invoiceNumber,
      parent_invoice_id:
        self.parentInvoiceId != null
          ? InvoiceIdSerializer._toJsonObject(self.parentInvoiceId)
          : undefined,
      status: InvoiceStatusSerializer._toJsonObject(self.status),
      tax_amount: self.taxAmount,
      total: self.total,
    };
  },
};
